odoo/addons/base/tests/test_ir_mail_server.py

```diff
@@ -0,0 +1,390 @@
+        self.assert_email_sent_smtp(
+            smtp_from=default_bounce_adress,
+            message_from='test_user@test.com',
+            from_filter='test.com',
+        )
+
+        # Use an email outside of the domain of the "from_filter"
+        # So we will use the notifications email in the headers and the bounce address
+        # in the envelop because the "from_filter" allows to use the entire domain
+        with self.mock_smtplib_connection():
+            message = self._build_email(mail_from='bob@mail.example.com')
+            IrMailServer.send_email(message)
+
+        self.connect_mocked.assert_called_once()
+        self.assert_email_sent_smtp(
+            smtp_from=default_bounce_adress,
+            message_from='"bob@mail.example.com" <notifications@test.com>',
+            from_filter='test.com',
+        )
+
+    @mute_logger('odoo.models.unlink')
+    @patch.dict("odoo.tools.config.options", {"from_filter": "test.com"})
+    def test_mail_server_binary_arguments_domain_smtp_session(self):
+        """Test the configuration provided in the odoo-bin arguments.
+
+        This config is used when no mail server exists.
+        Use a pre-configured SMTP session.
+        """
+        IrMailServer = self.env['ir.mail_server']
+        default_bounce_adress = self.env['ir.mail_server']._get_default_bounce_address()
+
+        # Remove all mail server so we will use the odoo-bin arguments
+        self.env['ir.mail_server'].search([]).unlink()
+        self.assertFalse(self.env['ir.mail_server'].search([]))
+
+        # Use an email in the domain of the "from_filter"
+        with self.mock_smtplib_connection():
+            smtp_session = IrMailServer.connect(smtp_from='test_user@test.com')
+            message = self._build_email(mail_from='test_user@test.com')
+            IrMailServer.send_email(message, smtp_session=smtp_session)
+
+        self.connect_mocked.assert_called_once()
+        self.assert_email_sent_smtp(
+            smtp_from=default_bounce_adress,
+            message_from='test_user@test.com',
+            from_filter='test.com',
+        )
+
+        # Use an email outside of the domain of the "from_filter"
+        # So we will use the notifications email in the headers and the bounce address
+        # in the envelop because the "from_filter" allows to use the entire domain
+        with self.mock_smtplib_connection():
+            smtp_session = IrMailServer.connect(smtp_from='bob@mail.example.com')
+            message = self._build_email(mail_from='bob@mail.example.com')
+            IrMailServer.send_email(message, smtp_session=smtp_session)
+
+        self.connect_mocked.assert_called_once()
+        self.assert_email_sent_smtp(
+            smtp_from=default_bounce_adress,
+            message_from='"bob@mail.example.com" <notifications@test.com>',
+            from_filter='test.com',
+        )
+
+    def _build_email(self, mail_from, return_path=None):
```

Comment:
Linting: would move it on top of class below setup/setupclass, to have setup methods,then tools methods, then all test functions.